Feature/external event building

a-d-smith opened the pull request:
Please see these slides for a detailed explanation of the changes made
https://www.hep.phy.cam.ac.uk/~asmith/29062018-ExternalEventBuilding.pdf

If you want to run this with MicroBooNE (or anything that doesn't use the standard settings files), there are a few additional lines that are required for the pandora neutrino settings files:

```xml
<!-- Track and shower building -->
<algorithm type = "LArTrackParticleBuilding">
    <PfoListName>NeutrinoParticles3D</PfoListName>
    <VertexListName>NeutrinoVertices3D</VertexListName>
</algorithm>
<algorithm type = "LArPcaShowerParticleBuilding">
    <PfoListName>NeutrinoParticles3D</PfoListName>
    <VertexListName>NeutrinoVertices3D</VertexListName>
</algorithm>
```

PandoraPFA commented on larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc:
```diff
@@ -0,0 +1,352 @@
+            // Check if the parent PFParticle is clear cosmic
+            const auto parentIterator(particlesToMetadata.find(LArPandoraHelper::GetParentPFParticle(particleMap, entry.first)));
+
+            if (static_cast<bool>(std::round(this->GetMetadataValue(parentIterator->second, "IsClearCosmic"))))
```
Great use of `std::round`!

PandoraPFA commented on larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc:
```diff
@@ -0,0 +1,352 @@
+    std::map<unsigned int, PFParticleVector> nuHypotheses;
+
+    // Collect the slice information
+    for (const auto &entry : particlesToMetadata)
```
Again check ordering is defined and reproducible.

PandoraPFA commented on larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc:
```diff
@@ -0,0 +1,352 @@
+            throw cet::exception("LArPandoraExternalEventBuilding") << "Can't find the parent of input PFParticle" << std::endl;
+        
+        const int pdg(std::abs(parentIterator->first->PdgCode())); 
+        const bool isNeutrino(std::abs(pdg) == pandora::NU_E || std::abs(pdg) == pandora::NU_MU || std::abs(pdg) == pandora::NU_TAU);
```
Check/use/update `LArPandoraHelper::IsNeutrino`

PandoraPFA commented on larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc:
```diff
@@ -0,0 +1,352 @@
+            // The above information should only unavailable for clear cosmic PFParticles
+            try
+            {
+                if (static_cast<bool>(std::round(this->GetMetadataValue(parentIterator->second, "IsClearCosmic")))) continue;
```
Maybe better to fully-resolve fallout from first exception before any further operations? Then move on to this treatment? I know you raise the same exception instance again, but don't much like much work inside a catch block.

Also , prefer not to have continue on the same line, but hey...

PandoraPFA commented on larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    // Produce the slices
+    for (unsigned int sliceId = 1; sliceId <= nuScores.size(); ++sliceId)
```
Here we see the "inside knowledge" that sliceIds start at 1... Might want to add an ATTN or similar.

PandoraPFA commented on larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc:
```diff
@@ -0,0 +1,352 @@
+    }
+
+    // ATTN This is required to ensure that the particles remain in order!
+    for (const auto &part : allParticles)
```
Maybe one more line to elaborate upon need for this?

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraOutput.cxx:
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
 
-    if (!settings.m_pPrimaryPandora)
-        throw cet::exception("LArPandora") << " LArPandoraOutput::ProduceArtOutput --- primary Pandora instance does not exist ";
+    // ---
```
If you need to break up implementation with ascii art, then refactor... :)

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraOutput.cxx:
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
-        throw cet::exception("LArPandora") << " LArPandoraOutput::ProduceArtOutput --- pointer to ART Producer module does not exist ";
+    // Collect immutable lists of pandora collections that we should convert to ART format
+    // The index of an object in a list will become the ID of the ART object produced
+    // Also obtain mappings between IDs in different collections which will become associations
```
This information would normally go in the header file, with the function prototype/definition.

a-d-smith replied:
This information relates to all of the `Collect*` functions, so I wasn't sure where to put it without just repeating.

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraOutput.cxx:
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
-    std::unique_ptr< std::vector<recob::Vertex> >                     outputVertices( new std::vector<recob::Vertex> );
-    std::unique_ptr< std::vector<anab::T0> >                          outputT0s( new std::vector<anab::T0> );
-    std::unique_ptr< std::vector<larpandoraobj::PFParticleMetadata> > outputParticleMetadata( new std::vector<larpandoraobj::PFParticleMetadata> );
+    // ---
```
Actually not looking too long, although beyond the 7+/-2 lines we should aspire too!!!

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraOutput.cxx:
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
+    for (const pandora::Pandora *const pPandora : MultiPandoraApi::GetDaughterPandoraInstanceList(pMasterPandora))
     {
-        particleMap.insert( std::pair<const pandora::ParticleFlowObject*, size_t>(pPfo, particleCounter++) );
+        const std::string name(pPandora->GetName());
```
const reference.

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraOutput.cxx:
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
+    // Identify the pandora worker instances by their name
+    const pandora::Pandora *pSlicingWorker(nullptr);
+    const pandora::Pandora *pSliceNuWorker(nullptr);
+    const pandora::Pandora *pSliceCRWorker(nullptr);
```
Obviously need to be super careful here for simple use cases where there are no slice workers, etc. and these are left holding `nullptr` values.

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraOutput.cxx:
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
-        ThreeDParticleMap::const_iterator iter = particleMap.find(pPfo);
-        if (particleMap.end() == iter)
-            throw cet::exception("LArPandora") << " LArPandoraOutput::ProduceArtOutput --- found an unassociated particle";
+    if (!pSlicingWorker || !pSliceNuWorker || !pSliceCRWorker)
```
What about other use cases? Do we just rely on setting `enableAllOutcomes` to false there, or else an exception will be raised quite quickly and terminate the job? Not a bad approach...

a-d-smith replied:
Yes, exactly that's the idea.

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraOutput.cxx:
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
+    PANDORA_THROW_RESULT_IF(pandora::STATUS_CODE_SUCCESS, !=, PandoraApi::GetCurrentPfoList(*pSlicingWorker, pSlicePfoList));
 
-        const size_t pfoIdCode(iter->second);
+    for (unsigned int sliceIndex = 0; sliceIndex < pSlicePfoList->size(); ++sliceIndex)
```
Here sliceIndex starts at 0. Are we being consistent throughout?

a-d-smith replied:
The PFO collection names start from zero, but the slices (as logged by Pandora) start from 1. Here I am just using the numbering system in each case from Pandora, but I can adopt one or the other if you prefer?

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraOutput.cxx:
```diff
@@ -47,269 +48,604 @@ namespace lar_pandora
-                throw cet::exception("LArPandora") << " LArPandoraOutput::ProduceArtOutput --- this particle has multiple parent particles ";
+            isClearCosmic = pPfo->GetPropertiesMap().at("IsClearCosmic");
+        }
+        catch (...)
```
Does this need to be `catch (...)`? Instances of which types may be raised?

a-d-smith commented on larpandora/LArPandoraObjects/classes_def.xml:
```diff
@@ -10,4 +10,10 @@
+    <class name="art::Wrapper<art::Assns<recob::PFParticle,bool,void> >"/>
+            
+    <class name="art::Assns<bool,recob::PFParticle,void>"/>
+    <class name="art::Wrapper<art::Assns<bool,recob::PFParticle,void> >"/>
```
Sorry, this was something I forgot to remove from an old iteration!

a-d-smith commented on larpandora/LArPandoraInterface/LArPandoraOutput.h:
```diff
@@ -7,15 +7,20 @@
 #include "larpandora/LArPandoraInterface/ILArPandora.h"
 #include "larpandora/LArPandoraInterface/LArPandoraHelper.h"
 
+#include "Pandora/PandoraInternal.h"
```
I used this to pick-up all the typedefs for `ClusterList`, `CaloHitVector`, etc... If preferable I can just make these typedefs here?

a-d-smith commented on larpandora/LArPandoraInterface/LArPandoraOutput.h:
```diff
@@ -50,46 +65,384 @@ class LArPandoraOutput
+     *  @param  idB the id of an object of type B to associate to the first object
+     *  @param  association the output association to update
+     */
+    template <typename A, typename B>
```
I guess this links up with your above comment about including `PandoraInternal.h`. Would you rather I didn't include that file?

PandoraPFA replied:
Well, a strict include "what you use" policy means you should include this file, but I tend to recommend that people just include Api/PandoraApi.h and it should give them everything they need to write a Pandora application.

This starts to get subjective, and there are arguments for and against. Ultimately include what you use would probably win, but just including the Api header would be clean.

a-d-smith commented on larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc:
```diff
@@ -0,0 +1,352 @@
+    void produce(art::Event &evt) override;
+
+private:
+    typedef std::map<art::Ptr<recob::PFParticle>, art::Ptr<larpandoraobj::PFParticleMetadata> > PFParticleToMetadata;
```
This is a new typedef, the nearest typedef in LArPandoraHelper is PFParticle to vector<Metadata>

a-d-smith commented on larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc:
```diff
@@ -0,0 +1,352 @@
+    m_trackProducerLabel(pset.get<std::string>("TrackProducerLabel")),
+    m_showerProducerLabel(pset.get<std::string>("ShowerProducerLabel")),
+    m_hitProducerLabel(pset.get<std::string>("HitProducerLabel")),
+    m_shouldProduceT0s(pset.get<bool>("ShouldProduceT0s", false)),
```
Looking it up, LArPandora actually has no default for ShouldRunStitching (which controls if T0s are produced). I will remove this default to be consistent

a-d-smith commented on larpandora/LArPandoraEventBuilding/LArPandoraExternalEventBuilding_module.cc:
```diff
@@ -0,0 +1,352 @@
+    std::string                         m_showerProducerLabel; ///< Label for the shower producer using the Pandora instance that produced the collections we want to consolidate
+    std::string                         m_hitProducerLabel;    ///< Label for the hit producer that was used as input to the Pandora instance specified
+    bool                                m_shouldProduceT0s;    ///< If we should produce T0s (relevant when stitching over multiple drift volumes)
+    art::InputTag                       m_pandoraTag;          ///< The input tag for the pandora producer
```
Me neither until recently, they are nifty!

a-d-smith commented:
Hello! 

I have hopefully addressed your comments so far as best I can. I have also - where easy to do so - used the LArPfoHelper instead of reinventing the wheel, and simplified some confusing try-catch logic. There are still some nastinesses lurking where I haven't changed the implementation of certain functions from the pre-factored code.

I will circulate some slides soon with instructions on how best to test. I think the event dumping module I wrote a while ago will be the most useful tool as it can be easily diffed! In my test so far, things seem to be working as expected but (as you say) much more testing is required for all the experiments this touches.

PandoraPFA commented:
Hi Andy,

Many thanks!

I think you should now distribute some simple slides with some clear tests indicated - ideally a concrete before/after here is what should/shouldn't change. Then you'll need to gently persuade :) everyone to do a bit of testing - I should have time too, especially if the instructions about e.g. the .fcl configuration for test jobs are nice and clear.

Once we're happy with that, I suggest that you and I sit down and read through LArPandoraOutput.cxx together again to double check, then we go for a talk and a release. Does that work for you?

John

StevenGreen1 commented:
Merged via the command line.

AndyChappell reviewed (commented):
The changes look logical and correct to me. Most of the suggested changes are cosmetic. I realise that for a lot of cases, this is just maintaining the existing style within the relevant file, which doesn't really conform to the usual Pandora standard. I think it makes sense to try to adopt the Pandora standard wherever reasonable with new code and then at some point in the future try to fix the overall formatting of these files. @PandoraPFA, what do you think?
